# Лабораторна робота №4  
## Тема: Рекурсія. Стратегія «розділяй і володарюй»
### **Студент:** Варакута Олександр 
### **Група:** КІ-24-1 

## Мета роботи: Засвоїти принципи побудови рекурсивних алгоритмів, реалізацію рекурсії мовою Python, а також навчитися оцінювати асимптотичну складність рекурсивних алгоритмів із використанням основної теореми рекурсії.

## Короткі теоретичні відомості

**Рекурсія** — це метод програмування, за якого функція викликає саму себе для розв’язання підзадач меншого розміру.

Кожен рекурсивний алгоритм обов’язково містить:
- **базовий випадок** — умову завершення рекурсії;
- **рекурсивний крок** — виклик функції з аргументом меншого розміру.

Для аналізу складності рекурсивних алгоритмів використовується рекурентне співвідношення виду:

$$
T(n) = aT\left(\frac{n}{b}\right) + O(n^d),
$$

де:
- $a$ — кількість підзадач,
- $b$ — коефіцієнт зменшення розміру задачі,
- $O(n^d)$ — складність об’єднання результатів.


## 1. Обчислення факторіалу


### 1.1 Обчислення факторіалу за допомогою циклу `for`

Нехай задано $n = 6$.

Факторіал визначається формулою:

$$
n! = 1 \cdot 2 \cdot 3 \cdot \dots \cdot n
$$


In [1]:
def factorial_for(n):
    result = 1
    for i in range(1, n + 1):
        result *= i
    return result

factorial_for(6)


720

**Асимптотична складність**

Алгоритм містить цикл, який виконується $n$ разів, і кожна ітерація виконується за сталий час.

$$
T(n) = O(n)
$$


### 1.2 Обчислення факторіалу рекурсивно

Нехай $n = 5$.

Рекурсивне означення факторіалу:

$$
n! =
\begin{cases}
1, & \text{якщо } n = 0, \\
n \cdot (n-1)!, & \text{якщо } n > 0.
\end{cases}
$$


In [2]:
def factorial_recursive(n):
    if n == 0:
        return 1
    return n * factorial_recursive(n - 1)

factorial_recursive(5)


120

Рекурсивні виклики:

$$
\begin{aligned}
5! &= 5 \cdot 4! \\
4! &= 4 \cdot 3! \\
3! &= 3 \cdot 2! \\
2! &= 2 \cdot 1! \\
1! &= 1 \cdot 0! = 1
\end{aligned}
$$

Результат: $5! = 120$.


**Асимптотична складність рекурсивного алгоритму**

Рекурентне співвідношення:

$$
T(n) = T(n-1) + O(1)
$$

Розв’язок:

$$
T(n) = O(n)
$$


## 2. Обчислення чисел Фібоначчі


Послідовність Фібоначчі визначається так:

$$
F_n =
\begin{cases}
0, & n = 0, \\
1, & n = 1, \\
F_{n-1} + F_{n-2}, & n \ge 2.
\end{cases}
$$

Обчислимо $F_8$.


In [3]:
def fibonacci(n):
    if n == 0:
        return 0
    if n == 1:
        return 1
    return fibonacci(n - 1) + fibonacci(n - 2)

fibonacci(8)


21

**Асимптотична складність**

Рекурентне співвідношення:

$$
T(n) = T(n-1) + T(n-2)
$$

Кількість рекурсивних викликів зростає експоненційно:

$$
T(n) = O(2^n)
$$

Такий алгоритм є неефективним для великих $n$.


## 3. Сортування злиттям (Merge Sort)


In [7]:
data = [34, 7, 23, 32, 5, 62, 15, 4]
def merge(left, right):
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    return result
def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)

merge_sort(data)



[4, 5, 7, 15, 23, 32, 34, 62]

**Асимптотична складність**

Рекурентне співвідношення:

$$
T(n) = 2T\left(\frac{n}{2}\right) + O(n)
$$

Згідно з теоремою рекурсії:

$$
T(n) = O(n \log n)
$$


## Контрольні питання
### 1. Що таке рекурсія?

Рекурсія — це метод програмування, при якому задача розв’язується шляхом зведення її до задач того самого типу меншого розміру. У програмуванні рекурсія реалізується шляхом виклику функцією самої себе до досягнення базового випадку.

### 2. Що таке базовий випадок рекурсії і чому він необхідний?

Базовий випадок — це умова, при якій рекурсивні виклики припиняються. Він необхідний для запобігання нескінченній рекурсії та переповненню стеку викликів. Наприклад, у факторіалі базовим випадком є $0! = 1$.

### 3. У чому полягає стратегія «розділяй і володарюй»?

Стратегія «розділяй і володарюй» полягає у поділі складної задачі на кілька простіших підзадач, рекурсивному розв’язанні цих підзадач та об’єднанні отриманих результатів у розв’язок початкової задачі.

### 4. Чому рекурсивне обчислення чисел Фібоначчі є неефективним?

Рекурсивний алгоритм Фібоначчі багаторазово обчислює одні й ті самі значення, що призводить до експоненційної кількості викликів і часової складності $O(2^n)$.

### 5. Яку асимптотичну складність має сортування злиттям?

Сортування злиттям має часову складність $O(n \log n)$ у найгіршому, середньому та найкращому випадках.

